### DISCUSSION TITLE
Is there Joomla4 compatibility?

### DISCUSSION THREAD

**Discussion by HiRolla** [joomlatools/joomlatools-pages#872] [Q&A]
Hello! I would like to deploy website using Pages. I've installed Joomlatools Framework and then installed Pages. But the Pages is not listed anywhere in Joomla/administrator. And the Quickstart example does not work for me: /index.php/hello returns 404.

**Comment by johanjanssens** [comment]
Pages doesn't work on Joomla 4 yet, to get started please use Joomla 3, once I have finished the compatibility for Joomla 4 you can then just copy your pages setup there and it will just work. 

It's perfectly normal you don't find he component in the administrator. Pages is not installed as a component, it's a developer library. It doesn't have any user interface.

You can find more info in the wiki how to get started.

1. Installation instructions: https://github.com/joomlatools/joomlatools-pages/wiki/Installation#installing-with-the-joomla-installer
2. Quick start: https://github.com/joomlatools/joomlatools-pages/wiki/Quick-Start

If /index.php/hello returns a 404, you either did something wrong, or you might have url rewriting enabled in which case you need to use /hello.

**Reply by HiRolla** [reply]
@johanjanssens ok, thanks! I will go back to Joomla 3 for a while.
Your projects are just amazing!

**Reply by HiRolla** [reply]
@johanjanssens I've followed the instructions, but still don't get the quickstart working on Joomla 3.10.5. Should I go to 3.9?

**Reply by HiRolla** [reply]
@johanjanssens It seems like I miss something, because Joomla 3.9 also gives me "404". 
1. Installed clean Joomla 3.10 with default settings
2. Installed joomlatools-framework-4.0.1.zip
3. Installed com_pages_v0.22.3.zip
4. Created /joomlatools-pages/pages/hello.html
5. Visited /index.php/hello, got 404
6. Visited /hello, got 404

Got the same result on Joomla 3.9
![image](https://user-images.githubusercontent.com/7746812/153258329-4d6e30d2-2530-439d-b9c1-98336b9cd289.png)

**Reply by johanjanssens** [reply]
You don't need to install Joomlatools Framework separately, the framework is included with the Pages package you install. When working with Pages always enable debug and set error reporting to max in your Joomla settings so you at least get proper errors when something goes wrong.

The error you show is a generic nginx error page, this is not generated by Joomla, it means that something is going wrong with /index.php/hello 

Steps:

1. First setup your Joomla site properly and setup SEF routing and url rewriting.
2. Check that the / home page of your site is working. 

Then go to any url that doesn't exit, like for example https://example.com/test, and see what you get, you should get a Joomla error page. If you don't then you have a server config issue and this problem is not related to pages, which based on the screenshot is quite likely.

**Reply by HiRolla** [reply]
@johanjanssens 
Done. First setup your Joomla site properly and setup SEF routing and url rewriting.
Done.  Check that the / home page of your site is working.
Done. https://mirovod.org/ Is working
Done. https://mirovod.org/test Gives 404 

https://mirovod.org/joomlatools-pages/pages/hello.html Exists
https://mirovod.org/hello = not found
https://mirovod.org/index.php/hello = not found

**Reply by johanjanssens** [reply]
Thanks, https://mirovod.org/test this is strange, could be because of you how setup Nginx rewrite rules. 

What you would expect here is the Joomla error page that the menu item or page doesn't exist, this error is rendered by Joomla, or when you have debug enabled and Pages installed you will get the Pages debugger to show up. 

Example: https://www.joomla.org/test

**Reply by HiRolla** [reply]
@johanjanssens I guess that my devault VPS config is not standard for Joomla and Pages. I did not change anything in it's configuration and I don't know what should I change. So I just switched from NGINX to Apache (FastCGI). Also I've switched off Joomla's "Use URL Rewriting" and finally I've got  some debug info on https://mirovod.org/index.php/hello 

In all other cases there was "404" from the webserver, not from Joomla.

Bot now I'm lost in this debug info. It seems like there is no Pages mentioned.

Upd: Couldn't setup on nginx, but on Apache (FastCGI) i've renamed htaccess.txt to .htaccess and finally got the Quickstart running. Thank you, @johanjanssens! I don't know if I should deploy my site on the slowest Apache (FastCGI), but I'm really glad to see this Hello world :)

**Reply by johanjanssens** [reply]
Glad to hear you got started. 

The issue with Nginx is not a Pages issue, but a problem with your Nginx setup and your rewrite rules. Best to check how to setup Joomla with Nginx, if Joomla works Pages will work fine too. 

About Apache, Apache isn't slow, on the contrary, even if you use Nginx you need to use FastCGI/PHP-FPM which becomes the bottleneck. In this scenario there is little or no difference using Nginx vs Apache in the content of Joomla. 

I personally prefer Apache. I find it easier to setup and configure, and offers more flexibility. I'm using it in [Joomlatools Server](https://github.com/joomlatools/joomlatools-server) and been very happy with the way it handles itself.

**Reply by HiRolla** [reply]
Thank you! Now I understand that Apache fits better in my case! Next time I'll take it instead of Nginx.

**Reply by johanjanssens** [reply]
To boost performance, I would make sure:

- you can properly configure the Pages http cache 
- combine it with a CDN proxy cache. (*)
-  enable the Pages Prefetcher: https://github.com/joomlatools/joomlatools-pages/pull/285 

(*) I personally prefer CloudFlare, it offers everything you need for free.  Related to CloudFlare you can read: https://github.com/joomlatools/joomlatools-pages/discussions/869#discussioncomment-2141846

You will get top notch performance, that is even faster than running a static site, for an idea of the results check https://www.joomlatools.com which is running this setup.

In a future step you can then considering using Joomlatools Server, this already works if you run Pages standalone, still working on making it run Joomla + Pages. You will then be able to benefit from the performance features (proxy cache, image optimisation, ...) that Joomlatools Server offers, more info: https://github.com/joomlatools/joomlatools-pages/discussions/849

**Comment by desbear** [comment]
Will joomlatools pages support joomla4 before joomla3.10 end of support ? 17 Aug 2023

**Reply by johanjanssens** [reply]
Still in on my to do yes.

**Reply by HiRolla** [reply]
Hello, @johanjanssens !
Joomlatools-Pages is a truly magnificent and irreplaceable invention! I can't stop thinking about it. Can we expect an update for Joomla 4?

**Comment by desbear** [comment]
Is there any workaround so that the joomlatools page can support joomla4?
